include/pmacc/math/Matrix.hpp

```diff
@@ -0,0 +1,201 @@
+/* Copyright 2023 Brian Marre, Rene Widera
+ *
+ * This file is part of PMacc.
+ *
+ * PMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with PMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//! @file implementation of matrix and matrix multiplication
+
+#pragma once
+
+#include "pmacc/attribute/unroll.hpp"
+#include "pmacc/math/Vector.hpp"
+#include "pmacc/memory/Array.hpp"
+#include "pmacc/static_assert.hpp"
+
+#include <cstdint>
+
+namespace pmacc::math
+{
+    /** 2 d matrix matrix definition
+     *
+     * @tparam T_Type data type used for storage of elements
+     * @tparam T_Extent pmacc::math::CT::Vector, tow dimensional vector describing extent of matrix in each dimensions
+     *
+     * @attention uses mathematical indexing
+     */
+    template<typename T_Type, typename T_Extent>
+    struct Matrix
+    {
+        PMACC_CASSERT_MSG(matrix_must_have_2_dimensional_extent, T_Extent::dim == 2);
+
+        using Extent = T_Extent;
+        using Type = T_Type;
+
+        static auto constexpr m = Extent::template at<0u>::type::value;
+        static auto constexpr n = Extent::template at<1u>::type::value;
+
+        using S_Elements = pmacc::memory::Array<T_Type, CT::volume<T_Extent>::type::value>;
+        using Idx = typename S_Elements::size_type;
+
+    private:
+        S_Elements elements;
+
+        /** get linear memory storage index from n-dimensional index
+         *
+         * indexation with the following scheme for a matrix with extent = < <N0>, <N1> >
+         *
+         * #(linear Index) | 0 | 1 | ... | <N0>-1 | <N0> | <N0> + 1 | ... | <N0>+<N0>-1 | 2 * <N0>| ...
+         * ----------------|---|---|-----|--------|------|----------|-----|-------------|---------|----
+         *                m| 0 | 1 | ... | <N0>-1 |    0 |        1 | ... |  (<N0> - 1) | 0, ...  | ...
+         *                n| 0 | 0 | ... |      0 |    1 |        1 | ... |           1 | 2, ...  | ...
+         *
+         * @attention no range checks outside debug, invalid input will lead to illegal memory access!
+         */
+        HDINLINE static Idx getLinearIndex(uint32_t const& mIdx, uint32_t const& nIdx)
```

Comment:
should be `HDINLINE static Idx getLinearIndex(uint32_t const mIdx, uint32_t const nIdx)`

Types with size `<=64` bit should be passed by copy, better for the compiler optimizations.

Reply:
Done!